extensions/stats/plugin.h

```diff
@@ -316,33 +151,115 @@ class SimpleStat {
+      MetricFactory{"response_bytes", MetricType::Histogram,
+
+                    [](const ::Wasm::Common::RequestInfo& request_info)
+                        -> uint64_t { return request_info.response_size; },
+                    false},
+      // TCP metrics.
+      MetricFactory{"tcp_sent_bytes_total", MetricType::Counter,
+                    [](const ::Wasm::Common::RequestInfo& request_info)
+                        -> uint64_t { return request_info.tcp_sent_bytes; },
+                    true},
+      MetricFactory{"tcp_received_bytes_total", MetricType::Counter,
+                    [](const ::Wasm::Common::RequestInfo& request_info)
+                        -> uint64_t { return request_info.tcp_received_bytes; },
+                    true},
+      MetricFactory{
+          "tcp_connections_opened_total", MetricType::Counter,
+          [](const ::Wasm::Common::RequestInfo& request_info) -> uint64_t {
+            return request_info.tcp_connections_opened;
+          },
+          true},
+      MetricFactory{
+          "tcp_connections_closed_total", MetricType::Counter,
+          [](const ::Wasm::Common::RequestInfo& request_info) -> uint64_t {
+            return request_info.tcp_connections_closed;
+          },
+          true},
+  };
+}
+
 // StatGen creates a SimpleStat based on resolved metric_id.
 class StatGen {
  public:
-  explicit StatGen(std::string name, MetricType metric_type,
+  explicit StatGen(const std::string& stat_prefix,
+                   const MetricFactory& metric_factory,
                    const std::vector<MetricTag>& tags,
-                   ValueExtractorFn value_fn, std::string field_separator,
-                   std::string value_separator, bool is_tcp_metric)
-      : name_(name),
-        value_fn_(value_fn),
-        metric_(metric_type, name, tags, field_separator, value_separator),
-        is_tcp_metric_(is_tcp_metric){};
+                   const std::vector<size_t>& indexes,
+                   const std::string& field_separator,
+                   const std::string& value_separator)
+      : is_tcp_(metric_factory.is_tcp),
+        indexes_(indexes),
+        extractor_(metric_factory.extractor),
+        metric_(metric_factory.type,
+                absl::StrCat(stat_prefix, metric_factory.name), tags,
+                field_separator, value_separator) {
+    if (tags.size() != indexes.size()) {
+      logAbort("metric tags.size() != indexes.size()");
+    }
+  };
 
   StatGen() = delete;
-  inline StringView name() const { return name_; };
-  inline bool is_tcp_metric() const { return is_tcp_metric_; }
+  inline StringView name() const { return metric_.name; };
+  inline bool is_tcp_metric() const { return is_tcp_; }
 
   // Resolve metric based on provided dimension values.
-  SimpleStat resolve(std::vector<std::string>& vals) {
-    auto metric_id = metric_.resolveWithFields(vals);
-    return SimpleStat(metric_id, value_fn_);
+  SimpleStat resolve(const IstioDimensions& instance) {
```

Comment:
Can we add comments to this function?

Reply:
Sure. The code is copy-pasted from ABI to avoid creating a temporary vector just to use the lower-level call.